From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 17 Oct 2022 16:50:37 +0200
Subject: [PATCH] Suki branding changes


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 994e81359154fc52392fe64cbe5f17a7a67e66c0..d77ed18ed4409bd7c6d0b42fea5d6e501cf2e4a4 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2415,8 +2415,10 @@ public final class Bukkit {
     public static long getLastTickMs() {
         return server.getLastTickMs();
     }
-    @NotNull public static java.time.Duration getLastTickTime() {
-        return server.getLastTickTime();
+    // Suki start - branding changes - Suki is not a fork of Yatopia
+    @NotNull public static java.time.Duration getLatestTickTime() {
+        return server.getLatestTickTime();
+        // Suki end - branding changes - Suki is not a fork of Yatopia
     }
     // Yatopia end
 
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index f25e6808c9ce7276c3e49c6ad03eda5ffa35e5f8..73a340b2488dcb3676354276b0acb39351839aea 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2116,7 +2116,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @Deprecated
     default long getLastTickMs() {
-        return getLastTickTime().toMillis();
+        return getLatestTickTime().toMillis(); // Suki - branding changes - Suki is not a fork of Yatopia
     }
     /**
      * Returns the time in {@link java.time.Duration} the last tick took.
@@ -2124,7 +2124,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return duration
      */
     @NotNull
-    java.time.Duration getLastTickTime();
+    java.time.Duration getLatestTickTime(); // Suki - branding changes - Suki is not a fork of Yatopia
     // Yatopia end
 
     /**
