From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 17 Oct 2022 16:50:37 +0200
Subject: [PATCH] Suki branding changes


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index bfa71be59dd84459bd420ace2d6694e1d80aee81..36e8d4a28211e81ad019ef8d8f9e1834a205aea7 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2435,8 +2435,10 @@ public final class Bukkit {
     public static long getLastTickMs() {
         return server.getLastTickMs();
     }
-    @NotNull public static java.time.Duration getLastTickTime() {
-        return server.getLastTickTime();
+    // Suki start - branding changes - Suki is not a fork of Yatopia
+    @NotNull public static java.time.Duration getLatestTickTime() {
+        return server.getLatestTickTime();
+        // Suki end - branding changes - Suki is not a fork of Yatopia
     }
     // Yatopia end
 
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 259209a999b80542a21ce6083301d3848d4006af..d3b1fe013576a63601043462c9750998bcf0866f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2137,7 +2137,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @Deprecated
     default long getLastTickMs() {
-        return getLastTickTime().toMillis();
+        return getLatestTickTime().toMillis(); // Suki - branding changes - Suki is not a fork of Yatopia
     }
     /**
      * Returns the time in {@link java.time.Duration} the last tick took.
@@ -2145,7 +2145,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return duration
      */
     @NotNull
-    java.time.Duration getLastTickTime();
+    java.time.Duration getLatestTickTime(); // Suki - branding changes - Suki is not a fork of Yatopia
     // Yatopia end
 
     /**
