From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 13 Oct 2022 20:55:09 +0200
Subject: [PATCH] Suki configuration and commands


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d3b1fe013576a63601043462c9750998bcf0866f..a9132e866a0a889eae8b36d369a211311cfa5ee5 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1967,6 +1967,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         {
             throw new UnsupportedOperationException("Not supported yet.");
         }
+
+        // Suki start - Suki configuration
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSukiConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Suki end - Suki configuration
+
         // Paper end
 
         // Mirai start
