From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 13 Oct 2022 20:55:09 +0200
Subject: [PATCH] Suki configuration and commands


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 73a340b2488dcb3676354276b0acb39351839aea..a55acc7b338f34bd7d8443ada50d25e4f9414967 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1946,6 +1946,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         {
             throw new UnsupportedOperationException("Not supported yet.");
         }
+
+        // Suki start - Suki configuration
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getSukiConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Suki end - Suki configuration
+
         // Paper end
 
         // Mirai start
