From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 22 Jun 2022 23:28:00 +0200
Subject: [PATCH] Custom stronghold generation


diff --git a/src/main/java/io/papermc/paper/world/structure/ConfiguredStructure.java b/src/main/java/io/papermc/paper/world/structure/ConfiguredStructure.java
index 5a43e40b7311ed2acb51f6ba8b12d1f34569ff2e..97810cbff0c4e8219a2d53823f40119ad13237cd 100644
--- a/src/main/java/io/papermc/paper/world/structure/ConfiguredStructure.java
+++ b/src/main/java/io/papermc/paper/world/structure/ConfiguredStructure.java
@@ -27,6 +27,7 @@ public final class ConfiguredStructure implements Keyed {
     public static final Reference<ConfiguredStructure> SHIPWRECK_BEACHED = create("shipwreck_beached");
     public static final Reference<ConfiguredStructure> SWAMP_HUT = create("swamp_hut");
     public static final Reference<ConfiguredStructure> STRONGHOLD = create("stronghold");
+    public static final Reference<ConfiguredStructure> STRONGHOLD_SURFACE = create("stronghold_surface"); // Martijn - custom strongholds - surface strongholds
     public static final Reference<ConfiguredStructure> OCEAN_MONUMENT = create("monument");
     public static final Reference<ConfiguredStructure> OCEAN_RUIN_COLD = create("ocean_ruin_cold");
     public static final Reference<ConfiguredStructure> OCEAN_RUIN_WARM = create("ocean_ruin_warm");
