From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 19 Jun 2022 23:44:33 +0200
Subject: [PATCH] More vanilla tags


diff --git a/src/main/java/org/bukkit/Tag.java b/src/main/java/org/bukkit/Tag.java
index 4749d4c01069f0e0cbe948ede1dd043baceaa97b..34fdab6ca0e0b4c1f70d9b849fcdc829f044166d 100644
--- a/src/main/java/org/bukkit/Tag.java
+++ b/src/main/java/org/bukkit/Tag.java
@@ -862,6 +862,31 @@ public interface Tag<T extends Keyed> extends Keyed {
     Tag<GameEvent> GAME_EVENT_IGNORE_VIBRATIONS_SNEAKING = Bukkit.getTag(REGISTRY_GAME_EVENTS, NamespacedKey.minecraft("ignore_vibrations_sneaking"), GameEvent.class);
     // Paper end
 
+    // Suki start - more vanilla tags
+
+    /**
+     * Vanilla item tag representing all swords.
+     */
+    Tag<Material> ITEMS_SWORDS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("swords"), Material.class);
+    /**
+     * Vanilla item tag representing all swords.
+     */
+    Tag<Material> ITEMS_HOES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("hoes"), Material.class);
+    /**
+     * Vanilla item tag representing all swords.
+     */
+    Tag<Material> ITEMS_PICKAXES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("pickaxes"), Material.class);
+    /**
+     * Vanilla item tag representing all swords.
+     */
+    Tag<Material> ITEMS_AXES = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("axes"), Material.class);
+    /**
+     * Vanilla item tag representing all swords.
+     */
+    Tag<Material> ITEMS_SHOVELS = Bukkit.getTag(REGISTRY_ITEMS, NamespacedKey.minecraft("shovels"), Material.class);
+
+    // Suki end - more vanilla tags
+
     /**
      * Returns whether or not this tag has an entry for the specified item.
      *
