From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 19 Jun 2022 00:13:36 +0200
Subject: [PATCH] Structure API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index c8c9addb2a5304e303556a9b53eb9ca1e7bd4f3a..cfb8595490fe2ce5edcf18973311363247f25c90 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1,6 +1,9 @@
 package org.bukkit;
 
 import java.io.File;
+
+import io.papermc.paper.world.structure.ConfiguredStructure;
+import org.sucraft.suki.structure.WorldStructure;
 import org.bukkit.generator.ChunkGenerator;
 
 import java.util.ArrayList;
@@ -75,6 +78,28 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     int getPlayerCount();
     // Paper end
 
+    // Suki start - structure API
+
+    /**
+     * Gets all structures at the given coordinates,
+     * with an optional type (if null, will get the structures of any type)
+     *
+     * @param block A block
+     * @param structureType An optional type of structure
+     * @return A list of structures found at the location
+     */
+    @NotNull List<@NotNull WorldStructure> getWorldStructuresAt(@NotNull Block block, @Nullable ConfiguredStructure structureType);
+
+    /**
+     * Gets all structures at the given coordinates
+     *
+     * @param block A block
+     * @return A list of structures found at the location
+     */
+    @NotNull List<@NotNull WorldStructure> getWorldStructuresAt(@NotNull Block block);
+
+    // Suki end - structure API
+
     /**
      * Gets the {@link Block} at the given coordinates
      *
diff --git a/src/main/java/org/sucraft/suki/structure/WorldStructure.java b/src/main/java/org/sucraft/suki/structure/WorldStructure.java
new file mode 100644
index 0000000000000000000000000000000000000000..b3a209dbaed700507775d109ac3c170a61f9b509
--- /dev/null
+++ b/src/main/java/org/sucraft/suki/structure/WorldStructure.java
@@ -0,0 +1,54 @@
+// Suki - structure API
+
+package org.sucraft.suki.structure;
+
+import io.papermc.paper.world.structure.ConfiguredStructure;
+import org.bukkit.StructureType;
+import org.bukkit.World;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+/**
+ * A structure somewhere in the world
+ * <br>
+ * Note that this structure may be invalid: meaning it has zero pieces
+ */
+public interface WorldStructure {
+
+    /**
+     * Gets the world that this structure is in
+     *
+     * @return The world that this structure is in
+     */
+    @NotNull World getWorld();
+
+    /**
+     * Gets whether this structure is valid
+     *
+     * @return Whether the structure is valid
+     */
+    boolean isValid();
+
+    /**
+     * Gets the configured type of this structure
+     *
+     * @return The configured type of this structure
+     */
+    @NotNull ConfiguredStructure getConfiguredStructure();
+
+    /**
+     * Gets the  type of this structure
+     *
+     * @return The type of this structure
+     */
+    @NotNull StructureType getStructureType();
+
+    /**
+     * Gets the pieces of this structure
+     *
+     * @return A list of the pieces
+     */
+    @NotNull List<@NotNull WorldStructurePiece> getPieces();
+
+}
diff --git a/src/main/java/org/sucraft/suki/structure/WorldStructureBoundingBox.java b/src/main/java/org/sucraft/suki/structure/WorldStructureBoundingBox.java
new file mode 100644
index 0000000000000000000000000000000000000000..b01aad79988db4df7c4b4a71d04d09432d51326b
--- /dev/null
+++ b/src/main/java/org/sucraft/suki/structure/WorldStructureBoundingBox.java
@@ -0,0 +1,9 @@
+// Suki - structure API
+
+package org.sucraft.suki.structure;
+
+public interface WorldStructureBoundingBox {
+
+
+
+}
diff --git a/src/main/java/org/sucraft/suki/structure/WorldStructurePiece.java b/src/main/java/org/sucraft/suki/structure/WorldStructurePiece.java
new file mode 100644
index 0000000000000000000000000000000000000000..9d359dfe7332a3a22d1e8f55620875276376201c
--- /dev/null
+++ b/src/main/java/org/sucraft/suki/structure/WorldStructurePiece.java
@@ -0,0 +1,64 @@
+// Suki - structure API
+
+package org.sucraft.suki.structure;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A piece of a structure somewhere in the world
+ */
+public interface WorldStructurePiece {
+
+    /**
+     * Gets the structure containing this piece
+     *
+     * @return The structure containing this piece
+     */
+    @NotNull WorldStructure getParent();
+
+    /**
+     * Gets the type of this structure piece
+     *
+     * @return The type of this structure piece
+     */
+    @NotNull WorldStructurePieceType getType();
+
+    /**
+     * Gets the template name of this piece's structure
+     *
+     * @return The template name of this piece's structure, or null if it is not present in this piece's NBT
+     */
+    @Nullable String getFullStructureTemplateName();
+
+    /**
+     * Gets the short template name of this piece's structure, which is the part of the full template name
+     * after the colon
+     *
+     * @return The short template name of this piece's structure, or null if it is not present in this piece's NBT
+     */
+    @Nullable String getShortStructureTemplateName();
+
+    /**
+     * Gets the pool element name of this structure piece
+     *
+     * @return The pool element name of this structure piece, or null if it is not present in this piece's NBT
+     */
+    @Nullable String getFullPoolElementName();
+
+    /**
+     * Gets the short pool element name of this structure piece, which is the part of the full pool element name
+     * after the colon
+     *
+     * @return The short pool element name of this structure piece, or null if it is not present in this piece's NBT
+     */
+    @Nullable String getShortPoolElementName();
+
+    /**
+     * Gets the bounding box of this structure piece
+     *
+     * @return The bounding box of this structure piece
+     */
+    @NotNull WorldStructureBoundingBox getBoundingBox();
+
+}
diff --git a/src/main/java/org/sucraft/suki/structure/WorldStructurePieceType.java b/src/main/java/org/sucraft/suki/structure/WorldStructurePieceType.java
new file mode 100644
index 0000000000000000000000000000000000000000..73540ebfcc32089f90adf1fa7824aa8ba0f8c4a7
--- /dev/null
+++ b/src/main/java/org/sucraft/suki/structure/WorldStructurePieceType.java
@@ -0,0 +1,8 @@
+// Suki - structure API
+
+package org.sucraft.suki.structure;
+
+/**
+ * A type of structure piece
+ */
+public interface WorldStructurePieceType {}
