From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 19 Sep 2022 22:49:15 +0200
Subject: [PATCH] Last tick time API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1be2129db92fde50885c53b830cc2c285509b501..8dfaead6ef64f199f49a027c8e6b5f3fa7f14d8e 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1096,7 +1096,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
     // Spigot End
 
     public static volatile RuntimeException chunkSystemCrash; // Paper - rewrite chunk system
-    public static java.time.Duration lastTickTime = java.time.Duration.ZERO; // Yatopia
+
+    // Suki start - last tick time API
+    public static java.time.Duration lastProperTickTime = java.time.Duration.ZERO;
+    public static java.time.Duration lastTickOversleepTime = java.time.Duration.ZERO;
+    // Suki end - last tick time API
 
     protected void runServer() {
         try {
@@ -1178,9 +1182,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
                 this.nextTickTime += 50L;
                 this.startMetricsRecordingTick();
                 this.profiler.push("tick");
-                long tickStart = System.nanoTime(); // Yatopia
+                long properTickStart = System.nanoTime(); // Suki - last tick time API
                 this.tickServer(this::haveTime);
-                lastTickTime = java.time.Duration.ofNanos(System.nanoTime() - tickStart); // Yatopia
+                lastProperTickTime = java.time.Duration.ofNanos(System.nanoTime() - properTickStart); // Suki - last tick time API
                 this.profiler.popPush("nextTickWait");
                 this.mayHaveDelayedTasks = true;
                 this.delayedTasksMaxNextTickTime = Math.max(Util.getMillis() + 50L, this.nextTickTime);
@@ -1296,9 +1300,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
 
     protected void waitUntilNextTick() {
         //this.executeAll(); // Paper - move this into the tick method for timings
+        long tickOversleepStart = System.nanoTime(); // Suki - last tick time API
         this.managedBlock(() -> {
             return !this.canSleepForTickNoOversleep(); // Paper - move oversleep into full server tick
         });
+        lastTickOversleepTime = java.time.Duration.ofNanos(System.nanoTime() - tickOversleepStart); // Suki - last tick time API
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f55cabbe49496d00fd9a7b63f4b4350beeb74edd..52b56b1d6f1a4b54bb63e9df38b7428baf9f8cbd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2923,10 +2923,18 @@ public final class CraftServer implements Server {
     }
     // Paper end
 
-    // Yatopia start
+    // Suki start - last tick time API
+
+    @Override
+    public java.time.Duration getLastProperTickTime() {
+        return MinecraftServer.lastProperTickTime;
+    }
+
     @Override
-    public java.time.Duration getLastTickTime() {
-        return net.minecraft.server.MinecraftServer.lastTickTime;
+    public java.time.Duration getLastTickOversleepTime() {
+        return MinecraftServer.lastTickOversleepTime;
     }
-    // Yatopia end
+
+    // Suki end - last tick time API
+
 }
diff --git a/src/main/java/org/spigotmc/TicksPerSecondCommand.java b/src/main/java/org/spigotmc/TicksPerSecondCommand.java
index 26ff7f68dacad945095d794c1def4752e0775abb..98daa177d87de6cccda7f2b4302799afdaddbf09 100644
--- a/src/main/java/org/spigotmc/TicksPerSecondCommand.java
+++ b/src/main/java/org/spigotmc/TicksPerSecondCommand.java
@@ -32,10 +32,6 @@ public class TicksPerSecondCommand extends Command
             tpsAvg[i] = TicksPerSecondCommand.format( tps[i] );
         }
         sender.sendMessage(ChatColor.GOLD + "TPS from last 5s, 1m, 5m, 15m: " + org.apache.commons.lang.StringUtils.join(tpsAvg, ", ")); // Purpur
-        // Yatopia start - Last tick time API
-        java.time.Duration lastTickTime = org.bukkit.Bukkit.getLastTickTime();
-        sender.sendMessage(ChatColor.GOLD + "Last tick: " + TicksPerSecondCommand.formatTo( lastTickTime, java.util.concurrent.TimeUnit.MILLISECONDS ) + " (" + formatTo( lastTickTime, java.util.concurrent.TimeUnit.NANOSECONDS ) + ")");
-        // Yatopia end
         if (args.length > 0 && args[0].equals("mem") && sender.hasPermission("bukkit.command.tpsmemory")) {
             sender.sendMessage(ChatColor.GOLD + "Current Memory Usage: " + ChatColor.GREEN + ((Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / (1024 * 1024)) + "/" + (Runtime.getRuntime().totalMemory() / (1024 * 1024)) + " mb (Max: " + (Runtime.getRuntime().maxMemory() / (1024 * 1024)) + " mb)");
             if (!hasShownMemoryWarning) {
