From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 13 Oct 2022 21:40:00 +0200
Subject: [PATCH] Do not log empty chat messages


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index c1f40592ee346e610d018542a65a79f2c9f5e8bd..8669862af25b53e295070b468211c963dfbd171e 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -243,6 +243,7 @@ import org.bukkit.inventory.CraftingInventory;
 import org.bukkit.inventory.EquipmentSlot;
 import org.bukkit.inventory.InventoryView;
 import org.bukkit.inventory.SmithingInventory;
+import org.sucraft.suki.configuration.SukiGlobalConfiguration;
 // CraftBukkit end
 
 public class ServerGamePacketListenerImpl implements ServerPlayerConnection, TickablePacketListener, ServerGamePacketListener {
@@ -2612,7 +2613,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         // CraftBukkit start
         String s = message.signedContent().plain();
         if (s.isEmpty()) {
-            ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried to send an empty message");
+            // Suki start - less console logs
+            if (SukiGlobalConfiguration.get().consoleLogs.chatMessages.empty) {
+                ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried to send an empty message");
+            }
+            // Suki end - less console logs
         } else if (this.getCraftPlayer().isConversing()) {
             OutgoingPlayerChatMessage outgoing = OutgoingPlayerChatMessage.create(message);
             outgoing.sendHeadersToRemainingPlayers(this.server.getPlayerList());
diff --git a/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java b/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java
index 1a6b8d4c8d7965b26affd54894862ff7a133f638..66a9af55e7c26bfd989c9142866d72193ab83792 100644
--- a/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java
+++ b/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java
@@ -25,7 +25,15 @@ public class SukiGlobalConfiguration extends ConfigurationPart {
     public ConsoleLogs consoleLogs;
 
     public class ConsoleLogs extends ConfigurationPart {
+
         public boolean invalidStatistics = true;
+        
+        public ChatMessages chatMessages;
+
+        public class ChatMessages extends ConfigurationPart {
+            public boolean empty = true;
+        }
+
     }
 
     // Suki end - less console logs
