From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 19 Jun 2022 23:44:17 +0200
Subject: [PATCH] More vanilla tags


diff --git a/src/main/java/net/minecraft/data/tags/ItemTagsProvider.java b/src/main/java/net/minecraft/data/tags/ItemTagsProvider.java
index ece5527231b3fa7d6022b37308f9f68fe563354b..2025d658534aca239794050cf09f958a015d3405 100644
--- a/src/main/java/net/minecraft/data/tags/ItemTagsProvider.java
+++ b/src/main/java/net/minecraft/data/tags/ItemTagsProvider.java
@@ -98,6 +98,17 @@ public class ItemTagsProvider extends TagsProvider<Item> {
         this.tag(ItemTags.AXOLOTL_TEMPT_ITEMS).add(Items.TROPICAL_FISH_BUCKET);
         this.tag(ItemTags.CLUSTER_MAX_HARVESTABLES).add(Items.DIAMOND_PICKAXE, Items.GOLDEN_PICKAXE, Items.IRON_PICKAXE, Items.NETHERITE_PICKAXE, Items.STONE_PICKAXE, Items.WOODEN_PICKAXE);
         this.tag(ItemTags.COMPASSES).add(Items.COMPASS).add(Items.RECOVERY_COMPASS);
+
+        // Martijn start - more vanilla tags
+
+        this.tag(ItemTags.SWORDS).add(Items.WOODEN_SWORD, Items.STONE_SWORD, Items.GOLDEN_SWORD, Items.IRON_SWORD, Items.DIAMOND_SWORD, Items.NETHERITE_SWORD);
+        this.tag(ItemTags.HOES).add(Items.WOODEN_HOE, Items.STONE_HOE, Items.GOLDEN_HOE, Items.IRON_HOE, Items.DIAMOND_HOE, Items.NETHERITE_HOE);
+        this.tag(ItemTags.PICKAXES).add(Items.WOODEN_PICKAXE, Items.STONE_PICKAXE, Items.GOLDEN_PICKAXE, Items.IRON_PICKAXE, Items.DIAMOND_PICKAXE, Items.NETHERITE_PICKAXE);
+        this.tag(ItemTags.AXES).add(Items.WOODEN_AXE, Items.STONE_AXE, Items.GOLDEN_AXE, Items.IRON_AXE, Items.DIAMOND_AXE, Items.NETHERITE_AXE);
+        this.tag(ItemTags.SHOVELS).add(Items.WOODEN_SHOVEL, Items.STONE_SHOVEL, Items.GOLDEN_SHOVEL, Items.IRON_SHOVEL, Items.DIAMOND_SHOVEL, Items.NETHERITE_SHOVEL);
+
+        // Martijn end - more vanilla tags
+        
     }
 
     protected void copy(TagKey<Block> blockTag, TagKey<Item> itemTag) {
diff --git a/src/main/java/net/minecraft/tags/ItemTags.java b/src/main/java/net/minecraft/tags/ItemTags.java
index e4a3e58e4523beb566063092bf09cf204915a08a..5404e3b8b7d096b76a6ddccf943f06adebac9487 100644
--- a/src/main/java/net/minecraft/tags/ItemTags.java
+++ b/src/main/java/net/minecraft/tags/ItemTags.java
@@ -83,6 +83,16 @@ public final class ItemTags {
     public static final TagKey<Item> CLUSTER_MAX_HARVESTABLES = bind("cluster_max_harvestables");
     public static final TagKey<Item> COMPASSES = bind("compasses");
 
+    // Martijn start - more vanilla tags
+
+    public static final TagKey<Item> SWORDS = bind("swords");
+    public static final TagKey<Item> HOES = bind("hoes");
+    public static final TagKey<Item> PICKAXES = bind("pickaxes");
+    public static final TagKey<Item> AXES = bind("axes");
+    public static final TagKey<Item> SHOVELS = bind("shovels");
+
+    // Martijn end - more vanilla tags
+
     private ItemTags() {
     }
 
