From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Fri, 10 Jun 2022 01:30:18 +0200
Subject: [PATCH] Hide plugins by permissions


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index cb79203603f5083a585b514115566ddad8a12668..2be6a210e276d5cf02dc487c721e7da1ac061419 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -169,6 +169,17 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     // Suki end - custom strongholds
 
+    // Suki start - hide plugins
+
+    public HidePlugins hidePlugins;
+
+    public class HidePlugins extends ConfigurationPart {
+        public boolean enabled = false;
+        public boolean hideLegacyAsterisk = false;
+    }
+
+    // Suki end - hide plugins
+
     public Messages messages;
 
     public class Messages extends ConfigurationPart {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index eefba0a6f2ed7bac0cba3463929b289d01cafc96..554da445f1827cb68022b5fdfe8ad69772a300e8 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -16,6 +16,7 @@ import com.mojang.serialization.Lifecycle;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufOutputStream;
 import io.netty.buffer.Unpooled;
+import io.papermc.paper.configuration.GlobalConfiguration;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import java.awt.image.BufferedImage;
 import java.io.File;
@@ -593,6 +594,18 @@ public final class CraftServer implements Server {
     }
     // Paper end
 
+    // Suki start - hide plugins
+    @Override
+    public boolean getHidePlugins() {
+        return GlobalConfiguration.get().hidePlugins.enabled;
+    }
+
+    @Override
+    public boolean getHideLegacyPluginAsterisk() {
+        return GlobalConfiguration.get().hidePlugins.hideLegacyAsterisk;
+    }
+    // Suki end - hide plugins
+
     @Override
     public List<CraftPlayer> getOnlinePlayers() {
         return this.playerView;
diff --git a/src/main/java/org/bukkit/craftbukkit/help/SimpleHelpMap.java b/src/main/java/org/bukkit/craftbukkit/help/SimpleHelpMap.java
index 40b66adcb5aac64212b1937dc506ebb60f2eed83..631d204bb9a320d757b04782154e2427240deed4 100644
--- a/src/main/java/org/bukkit/craftbukkit/help/SimpleHelpMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/help/SimpleHelpMap.java
@@ -11,6 +11,8 @@ import java.util.Map;
 import java.util.Set;
 import java.util.TreeMap;
 import java.util.TreeSet;
+
+import org.bukkit.Bukkit;
 import org.bukkit.command.Command;
 import org.bukkit.command.CommandExecutor;
 import org.bukkit.command.MultipleCommandAlias;
@@ -25,6 +27,7 @@ import org.bukkit.help.HelpTopic;
 import org.bukkit.help.HelpTopicComparator;
 import org.bukkit.help.HelpTopicFactory;
 import org.bukkit.help.IndexHelpTopic;
+import org.bukkit.plugin.PluginBase;
 
 /**
  * Standard implementation of {@link HelpMap} for CraftBukkit servers.
@@ -169,7 +172,7 @@ public class SimpleHelpMap implements HelpMap {
         this.fillPluginIndexes(pluginIndexes, this.server.getCommandMap().getCommands());
 
         for (Map.Entry<String, Set<HelpTopic>> entry : pluginIndexes.entrySet()) {
-            this.addTopic(new IndexHelpTopic(entry.getKey(), "All commands for " + entry.getKey(), null, entry.getValue(), "Below is a list of all " + entry.getKey() + " commands:"));
+            this.addTopic(new IndexHelpTopic(entry.getKey(), "All commands for " + entry.getKey(), Bukkit.getServer().getHidePlugins() ? PluginBase.getSeePluginPermissionName(entry.getKey()) : null, entry.getValue(), "Below is a list of all " + entry.getKey() + " commands:")); // Suki - hide plugins
         }
 
         // Amend help topics from the help.yml file
