From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 13 Oct 2022 21:40:00 +0200
Subject: [PATCH] Do not log empty chat messages


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 8b6a1cccd4e9071ef18904ccfa2669dc07cef3ae..29a959d2b65c988450e92ca36a3b483fa2550d10 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -244,6 +244,7 @@ import org.bukkit.inventory.CraftingInventory;
 import org.bukkit.inventory.EquipmentSlot;
 import org.bukkit.inventory.InventoryView;
 import org.bukkit.inventory.SmithingInventory;
+import org.sucraft.suki.configuration.SukiGlobalConfiguration;
 // CraftBukkit end
 
 public class ServerGamePacketListenerImpl implements ServerPlayerConnection, TickablePacketListener, ServerGamePacketListener {
@@ -2617,7 +2618,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         // CraftBukkit start
         String s = message.signedContent().plain();
         if (s.isEmpty()) {
-            ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried to send an empty message");
+            // Suki start - less console logs
+            if (SukiGlobalConfiguration.get().consoleLogs.chatMessages.empty) {
+                ServerGamePacketListenerImpl.LOGGER.warn(this.player.getScoreboardName() + " tried to send an empty message");
+            }
+            // Suki end - less console logs
         } else if (this.getCraftPlayer().isConversing()) {
             OutgoingPlayerChatMessage outgoing = OutgoingPlayerChatMessage.create(message);
             outgoing.sendHeadersToRemainingPlayers(this.server.getPlayerList());
diff --git a/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java b/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java
index 575c52f83bb39ef0fba85b02da1acda68b7f95aa..8ea32fb96d8d0303e3bac3d19cbf38bd88c7e69a 100644
--- a/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java
+++ b/src/main/java/org/sucraft/suki/configuration/SukiGlobalConfiguration.java
@@ -99,7 +99,15 @@ public class SukiGlobalConfiguration extends ConfigurationPart {
 
     public ConsoleLogs consoleLogs;
     public class ConsoleLogs extends ConfigurationPart {
+
         public boolean invalidStatistics = true;
+        
+        public ChatMessages chatMessages;
+
+        public class ChatMessages extends ConfigurationPart {
+            public boolean empty = true;
+        }
+
     }
 
     // Suki end - less console logs
