From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 19 Jan 2022 19:48:22 +0100
Subject: [PATCH] Zombies reinforcements on all difficulties


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index c91f109b4cf64dc1b4ef09f38e1cb8bf5cb2be13..588d88441bc5f01be97c9ef206a63ae56810b3a4 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -4,5 +4,4 @@ package com.destroystokyo.paper;
  * @deprecated kept as a means to identify Paper in older plugins/PaperLib
  */
 @Deprecated(forRemoval = true)
-public class PaperWorldConfig {
-}
+public class PaperWorldConfig {}
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index a9756c298348a0f1551841540bbe826732ae386a..9ba38eea00012cba6bd07876a059971033c9fa3c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -327,7 +327,7 @@ public class Zombie extends Monster {
                 entityliving = (LivingEntity) source.getEntity();
             }
 
-            if (entityliving != null && this.level.getDifficulty() == Difficulty.HARD && (double) this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE) && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
+            if (entityliving != null && (this.level.getDifficulty() == Difficulty.HARD || (this.level.getDifficulty() != Difficulty.PEACEFUL && this.level.sukiConfig().difficultyChanges.zombieReinforcementsOnAllDifficulties)) && (double) this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE) && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) { // Suki - zombies can call for reinforcements on all difficulties
                 int i = Mth.floor(this.getX());
                 int j = Mth.floor(this.getY());
                 int k = Mth.floor(this.getZ());
diff --git a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
index 0b3989191906237c71a85b5c0b7f746b736f0e7d..aca833af3e418d5a315b6ee343e434da7529df5e 100644
--- a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
+++ b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
@@ -171,4 +171,14 @@ public class SukiWorldConfiguration extends ConfigurationPart {
 
     // Suki end - breakable barrier blocks, breakable light blocks
 
+    // Suki start - difficulty changes
+
+    public DifficultyChanges difficultyChanges;
+
+    public class DifficultyChanges extends ConfigurationPart {
+        public boolean zombieReinforcementsOnAllDifficulties = false; // Suki - zombies can call for reinforcements on all difficulties
+    }
+
+    // Suki end - difficulty changes
+
 }
