From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Wed, 19 Jan 2022 19:48:22 +0100
Subject: [PATCH] Zombies can call for reinforcements on all difficulties


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index c91f109b4cf64dc1b4ef09f38e1cb8bf5cb2be13..588d88441bc5f01be97c9ef206a63ae56810b3a4 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -4,5 +4,4 @@ package com.destroystokyo.paper;
  * @deprecated kept as a means to identify Paper in older plugins/PaperLib
  */
 @Deprecated(forRemoval = true)
-public class PaperWorldConfig {
-}
+public class PaperWorldConfig {}
diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 503ec4de23cecc7f2d813a59703ea8165fa84c76..6cff7953d18f36728274d0c98ee1cfa28d2af39a 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -64,6 +64,16 @@ public class WorldConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     public int version = CURRENT_VERSION;
 
+    // Martijn start - difficulty changes
+
+    public DifficultyChanges difficultyChanges;
+
+    public class DifficultyChanges extends ConfigurationPart {
+        public boolean zombieReinforcementsOnAllDifficulties = false; // Martijn - zombies can call for reinforcements on all difficulties
+    }
+
+    // Martijn end - difficulty changes
+
     public Anticheat anticheat;
 
     public class Anticheat extends ConfigurationPart {
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index a9756c298348a0f1551841540bbe826732ae386a..1845cae0fabfe2bd3dff2453d936a4632dda2ad7 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -327,7 +327,7 @@ public class Zombie extends Monster {
                 entityliving = (LivingEntity) source.getEntity();
             }
 
-            if (entityliving != null && this.level.getDifficulty() == Difficulty.HARD && (double) this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE) && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
+            if (entityliving != null && (this.level.getDifficulty() == Difficulty.HARD || (this.level.getDifficulty() != Difficulty.PEACEFUL && this.level.paperConfig().difficultyChanges.zombieReinforcementsOnAllDifficulties)) && (double) this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE) && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) { // Martijn - zombies can call for reinforcements on all difficulties
                 int i = Mth.floor(this.getX());
                 int j = Mth.floor(this.getY());
                 int k = Mth.floor(this.getZ());
