From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 9 Jun 2022 01:37:34 +0200
Subject: [PATCH] Cave spiders can have poison on easy difficulty


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 24af73adecf255a0770fe54e3de841ba2a5a136d..1801ec672ddc661321fb594ea79e48ac5b4a12ff 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -71,6 +71,7 @@ public class WorldConfiguration extends ConfigurationPart {
     public class DifficultyChanges extends ConfigurationPart {
         public boolean zombieReinforcementsOnAllDifficulties = false; // Martijn - zombies can call for reinforcements on all difficulties
         public boolean raidsAreAlwaysHard = false; // Martijn - raids are always hard
+        public boolean caveSpidersHavePoisonOnEasyDifficulty = false; // Martijn - cave spiders have poison on easy difficulty
     }
 
     // Martijn end - difficulty changes
diff --git a/src/main/java/net/minecraft/world/entity/monster/CaveSpider.java b/src/main/java/net/minecraft/world/entity/monster/CaveSpider.java
index d980b906d9206560741576fa4153c57212f307a0..4c92c61b9c78a4f36ab30b6e8921728f641a27a1 100644
--- a/src/main/java/net/minecraft/world/entity/monster/CaveSpider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/CaveSpider.java
@@ -34,7 +34,7 @@ public class CaveSpider extends Spider {
             if (target instanceof LivingEntity) {
                 byte b0 = 0;
 
-                if (this.level.getDifficulty() == Difficulty.NORMAL) {
+                if (this.level.getDifficulty() == Difficulty.NORMAL || (this.level.getDifficulty() == Difficulty.EASY && this.level.paperConfig().difficultyChanges.caveSpidersHavePoisonOnEasyDifficulty)) { // Martijn - cave spiders have poison on easy difficulty
                     b0 = 7;
                 } else if (this.level.getDifficulty() == Difficulty.HARD) {
                     b0 = 15;
