From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 9 Jun 2022 01:59:25 +0200
Subject: [PATCH] Wither skulls can give the wither effect on easy difficulty


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 1801ec672ddc661321fb594ea79e48ac5b4a12ff..1eb8de139237a9b09a51355bb85d141a47a77ea3 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -72,6 +72,7 @@ public class WorldConfiguration extends ConfigurationPart {
         public boolean zombieReinforcementsOnAllDifficulties = false; // Suki - zombies can call for reinforcements on all difficulties
         public boolean raidsAreAlwaysHard = false; // Suki - raids are always hard
         public boolean caveSpidersHavePoisonOnEasyDifficulty = false; // Suki - cave spiders have poison on easy difficulty
+        public boolean witherEffectFromWitherSkullsOnEasyDifficulty = false; // Suki - wither effect from wither skulls on easy difficulty
     }
 
     // Suki end - difficulty changes
diff --git a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
index 2867e841e73a3edfdeb83af9d96e0d0cd4116a68..675de51295414d5b53458289ecda66d6dcd20593 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -75,7 +75,7 @@ public class WitherSkull extends AbstractHurtingProjectile {
             if (flag && entity instanceof LivingEntity) {
                 byte b0 = 0;
 
-                if (this.level.getDifficulty() == Difficulty.NORMAL) {
+                if (this.level.getDifficulty() == Difficulty.NORMAL || (this.level.getDifficulty() == Difficulty.EASY && this.level.paperConfig().difficultyChanges.witherEffectFromWitherSkullsOnEasyDifficulty)) { // Suki - wither effect from wither skulls on easy difficulty
                     b0 = 10;
                 } else if (this.level.getDifficulty() == Difficulty.HARD) {
                     b0 = 40;
