From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 22 Sep 2022 16:07:41 +0200
Subject: [PATCH] Players share villager gossips


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
index b6172cd6cadefa205c6b561ca2f4b88b1603f9ef..375c3d0dc3e8e8aab3b74ac0a87dac97059a9ec0 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
@@ -39,7 +39,7 @@ public class DefendVillageTargetGoal extends TargetGoal {
 
             while (iterator1.hasNext()) {
                 Player entityhuman = (Player) iterator1.next();
-                int i = entityvillager.getPlayerReputation(entityhuman);
+                int i = entityvillager.getPlayerReputation(entityhuman, false); // Suki - players share gossips
 
                 if (i <= -100) {
                     this.potentialTarget = entityhuman;
diff --git a/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java b/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java
index aa277479f5552503a202a057b1a3ede379f2bbbf..292ee703c477e68a6a929b1bc16f23ad4b5b23d7 100644
--- a/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java
+++ b/src/main/java/net/minecraft/world/entity/ai/gossip/GossipContainer.java
@@ -17,6 +17,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
+import java.util.function.BiPredicate;
 import java.util.function.DoublePredicate;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
@@ -24,6 +25,7 @@ import java.util.stream.Stream;
 import net.minecraft.core.UUIDUtil;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.VisibleForDebug;
+import org.jetbrains.annotations.Nullable;
 
 public class GossipContainer {
     public static final int DISCARD_THRESHOLD = 2;
@@ -114,7 +116,31 @@ public class GossipContainer {
         });
     }
 
-    public int getReputation(UUID target, Predicate<GossipType> gossipTypeFilter) {
+    // Suki start - players share gossips
+    public int getReputation(UUID target, Predicate<GossipType> gossipTypeFilter, @Nullable BiPredicate<UUID, GossipType> canUseGossipFromPlayer) {
+        if (canUseGossipFromPlayer != null) {
+            // Based on EntityGossips.weightedValue
+            int weight = 0;
+            for (GossipType gossipType : GossipType.values()) {
+//                @Nullable Integer bestWeightedValue = null;
+                if (gossipTypeFilter.test(gossipType)) {
+                    for (Map.Entry<UUID, EntityGossips> gossipsEntry : this.gossips.entrySet()) {
+                        if (canUseGossipFromPlayer.test(gossipsEntry.getKey(), gossipType)) {
+                            int weightedValue = gossipsEntry.getValue().entries.getInt(gossipType) * gossipType.weight;
+//                            if (bestWeightedValue == null || weightedValue > bestWeightedValue) {
+//                                bestWeightedValue = weightedValue;
+//                            }
+                            weight += weightedValue; // Using sum instead of max to stack curing events from different players
+                        }
+                    }
+                }
+//                if (bestWeightedValue != null) {
+//                    weight += bestWeightedValue;
+//                }
+            }
+            return weight;
+        }
+        // Suki end - players share gossips
         GossipContainer.EntityGossips entityGossips = this.gossips.get(target);
         return entityGossips != null ? entityGossips.weightedValue(gossipTypeFilter) : 0;
     }
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 3ea11bac3865102dab0ade99dacf8a54b9536c2e..c9f5f416782da2c1010fcc2af32fc41ec92099a7 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -9,6 +9,7 @@ import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
@@ -18,6 +19,7 @@ import java.util.Set;
 import java.util.function.BiPredicate;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.GlobalPos;
 import net.minecraft.core.Holder;
@@ -507,7 +509,7 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
     }
 
     private void updateSpecialPrices(Player player) {
-        int i = this.getPlayerReputation(player);
+        int i = this.getPlayerReputation(player, true); // Suki - players share gossips
 
         if (i != 0) {
             Iterator iterator = this.getOffers().iterator();
@@ -783,10 +785,15 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         }
     }
 
-    public int getPlayerReputation(Player player) {
+    // Suki start - players share gossips
+    public int getPlayerReputation(Player player, boolean checkSharedGossips) {
         return this.gossips.getReputation(player.getUUID(), (reputationtype) -> {
             return true;
-        });
+        }, checkSharedGossips ? ((otherPlayerUUID, gossipType) ->
+            ((gossipType == GossipType.MAJOR_POSITIVE || gossipType == GossipType.MINOR_POSITIVE) && this.level.sukiConfig().difficultyChanges.playersShareVillagerGossips.positive.always) ||
+                otherPlayerUUID.equals(player.getUUID())
+        ) : null);
+        // Suki end - players share gossips
     }
 
     private void digestFood(int amount) {
diff --git a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
index 2a4162887f9fefdfa2747724fb7cdd3b32a97bd9..86ad443f3fe003964111014de3be4df045382152 100644
--- a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
+++ b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
@@ -240,6 +240,24 @@ public class SukiWorldConfiguration extends ConfigurationPart {
 
         // Suki end - configurable villager reputation changes
 
+        // Suki start - players share gossips
+
+        public PlayersShareVillagerGossips playersShareVillagerGossips;
+
+        public class PlayersShareVillagerGossips extends ConfigurationPart {
+
+            public Positive positive;
+
+            public class Positive extends ConfigurationPart {
+
+                public boolean always = false;
+
+            }
+
+        }
+
+        // Suki end - players share gossips
+
     }
 
     // Suki end - difficulty changes
