From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 9 Jun 2022 02:09:22 +0200
Subject: [PATCH] Minimum health after starving


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index e219265fbab61e40028d784920b917fd5e09b353..b3e2964b7036a2fcfe4b92ae28601385ddec53f3 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -69,11 +69,23 @@ public class WorldConfiguration extends ConfigurationPart {
     public DifficultyChanges difficultyChanges;
 
     public class DifficultyChanges extends ConfigurationPart {
+
         public boolean zombieReinforcementsOnAllDifficulties = false; // Suki - zombies can call for reinforcements on all difficulties
         public boolean raidsAreAlwaysHard = false; // Suki - raids are always hard
         public boolean caveSpidersHavePoisonOnEasyDifficulty = false; // Suki - cave spiders have poison on easy difficulty
         public boolean witherEffectFromWitherSkullsOnEasyDifficulty = false; // Suki - wither effect from wither skulls on easy difficulty
         public boolean lightningSpawnsFire = true; // Suki - configurable lightning fire
+
+        // Suki start - minimum health after starving
+
+        /**
+         * -1 indicates no minimum (it may still be limited by the difficulty)
+         * Valid values are -1 or a nonnegative floating-point number
+         */
+        public float minimumHealthAfterStarving = -1f;
+
+        // Suki end - minimum health after starving
+
     }
 
     // Suki end - difficulty changes
diff --git a/src/main/java/net/minecraft/world/food/FoodData.java b/src/main/java/net/minecraft/world/food/FoodData.java
index 2934b6de1f1fb914a532ee20184df99d1acd8e65..b58d06d651729d7fbcc93ee1368e1bb6364a9c45 100644
--- a/src/main/java/net/minecraft/world/food/FoodData.java
+++ b/src/main/java/net/minecraft/world/food/FoodData.java
@@ -100,7 +100,7 @@ public class FoodData {
         } else if (this.foodLevel <= 0) {
             ++this.tickTimer;
             if (this.tickTimer >= this.starvationRate) { // CraftBukkit - add regen rate manipulation
-                if (player.getHealth() > 10.0F || enumdifficulty == Difficulty.HARD || player.getHealth() > 1.0F && enumdifficulty == Difficulty.NORMAL) {
+                if ((player.level.paperConfig().difficultyChanges.minimumHealthAfterStarving < 0.0F || player.getHealth() > player.level.paperConfig().difficultyChanges.minimumHealthAfterStarving) && (player.getHealth() > 10.0F || enumdifficulty == Difficulty.HARD || player.getHealth() > 1.0F && enumdifficulty == Difficulty.NORMAL)) { // Suki - minimum health after starving
                     player.hurt(DamageSource.STARVE, 1.0F);
                 }
 
