From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Thu, 9 Jun 2022 02:18:14 +0200
Subject: [PATCH] Spiders can spawn with effects on all difficulties


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index b3e2964b7036a2fcfe4b92ae28601385ddec53f3..fff8974ea2de60432bc3d9c9e00fa60294c31587 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -75,6 +75,7 @@ public class WorldConfiguration extends ConfigurationPart {
         public boolean caveSpidersHavePoisonOnEasyDifficulty = false; // Suki - cave spiders have poison on easy difficulty
         public boolean witherEffectFromWitherSkullsOnEasyDifficulty = false; // Suki - wither effect from wither skulls on easy difficulty
         public boolean lightningSpawnsFire = true; // Suki - configurable lightning fire
+        public boolean spidersWithEffectsOnAllDifficulties = false; // Suki - spiders can spawn with effects on all difficulties
 
         // Suki start - minimum health after starving
 
diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index c3085ad5410b41d9c7703b28ca835dbee3f49ee7..a007974ef1fdd9e11559b2c572dcd0b4cfed902a 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -168,7 +168,7 @@ public class Spider extends Monster {
 
         if (object == null) {
             object = new Spider.SpiderEffectsGroupData();
-            if (world.getDifficulty() == Difficulty.HARD && randomsource.nextFloat() < 0.1F * difficulty.getSpecialMultiplier()) {
+            if (((world.getDifficulty() != Difficulty.PEACEFUL && world.getLevel().paperConfig().difficultyChanges.spidersWithEffectsOnAllDifficulties) || world.getDifficulty() == Difficulty.HARD) && randomsource.nextFloat() < 0.1F * difficulty.getSpecialMultiplier()) { // Suki - spiders can spawn with effects on all difficulties
                 ((Spider.SpiderEffectsGroupData) object).setRandomEffect(randomsource);
             }
         }
