From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 20 Jun 2022 15:39:21 +0200
Subject: [PATCH] Variable entity wake-up duration


diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index 0153b40cef8b07660eb20e3a0dcb4fb19a1d44c8..4670d6a4d61f53f689efeb7666ecde85571fc8eb 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -60,9 +60,11 @@ import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.AABB;
 // Pufferfish start
 import net.minecraft.world.phys.Vec3;
+import java.util.Random;
 // Pufferfish end
 import org.bukkit.entity.Enderman;
 import org.bukkit.entity.Fish;
+import org.sucraft.suki.configuration.SukiWorldConfiguration;
 
 public class ActivationRange
 {
@@ -88,6 +90,20 @@ public class ActivationRange
         Activity.PANIC
     };
 
+    // Suki start - variable entity wake-up duration
+
+    private static Random wakeUpDurationRandom = new Random();
+
+    private static int getWakeUpDurationWithVariance(Entity entity, int wakeUpDuration) {
+        SukiWorldConfiguration config = entity.level.sukiConfig();
+        if (config.entityTicking.wakeUpDurationStandardDeviationRatio <= 0) {
+            return wakeUpDuration;
+        }
+        return (int) Math.min(Integer.MAX_VALUE, Math.max(1, Math.round(wakeUpDuration * wakeUpDurationRandom.nextGaussian(1, config.entityTicking.wakeUpDurationStandardDeviationRatio))));
+    }
+
+    // Suki end - variable entity wake-up duration
+
     private static int checkInactiveWakeup(Entity entity) {
         Level world = entity.level;
         SpigotWorldConfig config = world.spigotConfig;
@@ -95,22 +111,22 @@ public class ActivationRange
         if (entity.activationType == ActivationType.VILLAGER) {
             if (inactiveFor > config.wakeUpInactiveVillagersEvery && world.wakeupInactiveRemainingVillagers > 0) {
                 world.wakeupInactiveRemainingVillagers--;
-                return config.wakeUpInactiveVillagersFor;
+                return getWakeUpDurationWithVariance(entity, config.wakeUpInactiveVillagersFor); // Suki - variable entity wake-up duration
             }
         } else if (entity.activationType == ActivationType.ANIMAL) {
             if (inactiveFor > config.wakeUpInactiveAnimalsEvery && world.wakeupInactiveRemainingAnimals > 0) {
                 world.wakeupInactiveRemainingAnimals--;
-                return config.wakeUpInactiveAnimalsFor;
+                return getWakeUpDurationWithVariance(entity, config.wakeUpInactiveAnimalsFor); // Suki - variable entity wake-up duration
             }
         } else if (entity.activationType == ActivationType.FLYING_MONSTER) {
             if (inactiveFor > config.wakeUpInactiveFlyingEvery && world.wakeupInactiveRemainingFlying > 0) {
                 world.wakeupInactiveRemainingFlying--;
-                return config.wakeUpInactiveFlyingFor;
+                return getWakeUpDurationWithVariance(entity, config.wakeUpInactiveFlyingFor); // Suki - variable entity wake-up duration
             }
         } else if (entity.activationType == ActivationType.MONSTER || entity.activationType == ActivationType.RAIDER) {
             if (inactiveFor > config.wakeUpInactiveMonstersEvery && world.wakeupInactiveRemainingMonsters > 0) {
                 world.wakeupInactiveRemainingMonsters--;
-                return config.wakeUpInactiveMonstersFor;
+                return getWakeUpDurationWithVariance(entity, config.wakeUpInactiveMonstersFor); // Suki - variable entity wake-up duration
             }
         }
         return -1;
diff --git a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
index 0c88e2b03f951340293cceeba3a4c2a789b92d8c..b423c6df22aa995517234164b4222bedce4c2845 100644
--- a/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
+++ b/src/main/java/org/sucraft/suki/configuration/SukiWorldConfiguration.java
@@ -387,6 +387,16 @@ public class SukiWorldConfiguration extends ConfigurationPart {
 
         }
 
+        // Suki start - variable entity wake-up duration
+        /**
+         * This value is sigma,
+         * where the normal wake-up time will be multiplied by normal(mu = 1, sigma)
+         * <br>
+         * If this value is nonpositive, this feature will be disabled
+         */
+        public double wakeUpDurationStandardDeviationRatio = 0;
+        // Suki end - variable entity wake-up duration
+
     }
 
     // Suki end - more entity activation configuration
