From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 17 Jan 2022 12:37:38 +0100
Subject: [PATCH] Utility methods for entity-chunk distance


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 326c9fbad6bd7b63051a49feed1d80b6e3a3f575..436f6f860c7cb6dfad2e3176e527dd23591f0e0b 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -636,6 +636,36 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     // Suki end - utility functions for getting the entity's chunk
 
+    // Suki start - utility functions for getting distance to chunks
+
+    public int getChessboardDistanceToChunk(ChunkPos chunk) {
+        return Math.max(Math.abs(this.getChunkX() - chunk.x), Math.abs(this.getChunkZ() - chunk.z));
+    }
+
+    public int getChessboardDistanceToChunk(long chunkKey) {
+        return this.getChessboardDistanceToChunk(new ChunkPos(chunkKey));
+    };
+
+    public long getSquaredPythagoreanDistanceToChunk(ChunkPos chunk) {
+        long dx = this.getChunkX() - chunk.x;
+        long dz = this.getChunkZ() - chunk.z;
+        return dx * dx + dz * dz;
+    }
+
+    public long getSquaredPythagoreanDistanceToChunk(long chunkKey) {
+        return this.getSquaredPythagoreanDistanceToChunk(new ChunkPos(chunkKey));
+    };
+
+    public double getPythagoreanDistanceToChunk(ChunkPos chunkPos) {
+        return Math.sqrt(this.getSquaredPythagoreanDistanceToChunk(chunkPos));
+    }
+
+    public double getPythagoreanDistanceToChunk(long chunkKey) {
+        return this.getPythagoreanDistanceToChunk(new ChunkPos(chunkKey));
+    };
+
+    // Suki end - utility functions for getting distance to chunks
+
     public boolean isColliding(BlockPos pos, BlockState state) {
         VoxelShape voxelshape = state.getCollisionShape(this.level, pos, CollisionContext.of(this));
         VoxelShape voxelshape1 = voxelshape.move((double) pos.getX(), (double) pos.getY(), (double) pos.getZ());
