From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Tue, 21 Jun 2022 12:46:55 +0200
Subject: [PATCH] Deobfuscation helpers


diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 98664c95331cee4139711c402dfaf406ee672c22..2e2c199a4f43196758b228b72421f707e55a080d 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -510,6 +510,14 @@ public class Commands {
         return this.dispatcher;
     }
 
+    // Martijn start - OBFHELPER
+
+    public com.mojang.brigadier.CommandDispatcher<CommandSourceStack> _getDispatcher() {
+        return this.getDispatcher();
+    }
+
+    // Martijn end - OBFHELPER
+
     @Nullable
     public static <S> CommandSyntaxException getParseException(ParseResults<S> parse) {
         return !parse.getReader().canRead() ? null : (parse.getExceptions().size() == 1 ? (CommandSyntaxException) parse.getExceptions().values().iterator().next() : (parse.getContext().getRange().isEmpty() ? CommandSyntaxException.BUILT_IN_EXCEPTIONS.dispatcherUnknownCommand().createWithContext(parse.getReader()) : CommandSyntaxException.BUILT_IN_EXCEPTIONS.dispatcherUnknownArgument().createWithContext(parse.getReader())));
diff --git a/src/main/java/net/minecraft/nbt/Tag.java b/src/main/java/net/minecraft/nbt/Tag.java
index 2452fb49ebe5ec791cc7151e3b73b5c572ff2eda..c83e69a4a50dd69c16a7c08923170afb5daeb42a 100644
--- a/src/main/java/net/minecraft/nbt/Tag.java
+++ b/src/main/java/net/minecraft/nbt/Tag.java
@@ -39,6 +39,14 @@ public interface Tag {
         return (new StringTagVisitor()).visit(this);
     }
 
+    // Martijn start - OBFHELPER
+
+    default String _getAsString() {
+        return this.getAsString();
+    }
+
+    // Martijn end - OBFHELPER
+
     void accept(TagVisitor visitor);
 
     StreamTagVisitor.ValueResult accept(StreamTagVisitor visitor);
diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 8f3dfda2483885dad01edb444021ba3407e4896e..4ce2c89cfe7d9255ade257c2a1e6507f28ab2acc 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -76,6 +76,11 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     private final PacketFlow receiving;
     private final Queue<Connection.PacketHolder> queue = Queues.newConcurrentLinkedQueue();
     public Channel channel;
+    // Martijn start - OBFHELPER
+    public Channel _channel() {
+        return this.channel;
+    }
+    // Martijn end - OBFHELPER
     public SocketAddress address;
     // Spigot Start
     public java.util.UUID spoofedUUID;
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundGameEventPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundGameEventPacket.java
index 7a225421d54f9b30f0c62036888e09379ee28bd2..fe5d0bdf4c2e7efe3d4974ae106e091a42e157aa 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundGameEventPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundGameEventPacket.java
@@ -10,6 +10,11 @@ public class ClientboundGameEventPacket implements Packet<ClientGamePacketListen
     public static final ClientboundGameEventPacket.Type START_RAINING = new ClientboundGameEventPacket.Type(1);
     public static final ClientboundGameEventPacket.Type STOP_RAINING = new ClientboundGameEventPacket.Type(2);
     public static final ClientboundGameEventPacket.Type CHANGE_GAME_MODE = new ClientboundGameEventPacket.Type(3);
+    // Martijn start - OBFHELPER
+    public static ClientboundGameEventPacket.Type _CHANGE_GAME_MODE() {
+        return ClientboundGameEventPacket.CHANGE_GAME_MODE;
+    }
+    // Martijn end - OBFHELPER
     public static final ClientboundGameEventPacket.Type WIN_GAME = new ClientboundGameEventPacket.Type(4);
     public static final ClientboundGameEventPacket.Type DEMO_EVENT = new ClientboundGameEventPacket.Type(5);
     public static final ClientboundGameEventPacket.Type ARROW_HIT_PLAYER = new ClientboundGameEventPacket.Type(6);
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
index db0b69f227ab388fa1335471313200f072cca2f3..cd33f495c9fa98f106ff391ae2b0a13462736745 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -140,6 +140,14 @@ public class DedicatedPlayerList extends PlayerList {
         return (DedicatedServer)super.getServer();
     }
 
+    // Martijn start - OBFHELPER
+
+    public DedicatedServer _getServer() {
+        return this.getServer();
+    }
+
+    // Martijn end - OBFHELPER
+
     @Override
     public boolean canBypassPlayerLimit(GameProfile profile) {
         return this.getOps().canBypassPlayerLimit(profile);
diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 2d149e79ad180c12544de81f850fe713adfc8155..a7a7ad04e0ca9a9119075df7afe2ad5ba9d1695a 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -173,6 +173,13 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
     // Martijn start- per-player view distance
     // This value is the intended world send view distance + 1
     private int vanillaWorldViewDistancePlusOne;
+    // Martijn start - OBFHELPER
+    // This isn't actually obfuscated but we normally would want it to be private,
+    // but this opens the value up for plugins to use more readily when this is really the value that is needed
+    public int _vanillaWorldViewDistancePlusOne() {
+        return this.vanillaWorldViewDistancePlusOne;
+    }
+    // Martijn end - OBFHELPER
     // Martijn end - per-player view distance
     public final com.destroystokyo.paper.util.misc.PlayerAreaMap playerMobDistanceMap; // Paper
     public final ReferenceOpenHashSet<ChunkHolder> needsChangeBroadcasting = new ReferenceOpenHashSet<>();
diff --git a/src/main/java/net/minecraft/server/level/DistanceManager.java b/src/main/java/net/minecraft/server/level/DistanceManager.java
index 7922ced4e86d49df02169ac820fc5cbde974a18b..67e0257b98d0075b51171376eb2716afd5367e20 100644
--- a/src/main/java/net/minecraft/server/level/DistanceManager.java
+++ b/src/main/java/net/minecraft/server/level/DistanceManager.java
@@ -70,6 +70,11 @@ public abstract class DistanceManager {
     final Executor mainThreadExecutor;
     private long ticketTickCounter;
     private int simulationDistance = 10;
+    // Martijn start - OBFHELPER
+    public int _simulationDistance() {
+        return this.simulationDistance;
+    }
+    // Martijn end - OBFHELPER
     private final ChunkMap chunkMap; // Paper
 
     protected DistanceManager(Executor workerExecutor, Executor mainThreadExecutor, ChunkMap chunkMap) {
diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index 99b9d4951fadb8f6da800f948a841c9daa8e11c5..90cba9df54f9d5ac7f5e0f3a19e80a894483d7d8 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -54,11 +54,21 @@ public class ServerChunkCache extends ChunkSource {
     public static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger(); // Paper
     public static final List<ChunkStatus> CHUNK_STATUSES = ChunkStatus.getStatusList();
     private final DistanceManager distanceManager;
+    // Martijn start - OBFHELPER
+    public DistanceManager _distanceManager() {
+        return this.distanceManager;
+    }
+    // Martijn end - OBFHELPER
     final ServerLevel level;
     public final Thread mainThread;
     final ThreadedLevelLightEngine lightEngine;
     public final ServerChunkCache.MainThreadExecutor mainThreadProcessor;
     public final ChunkMap chunkMap;
+    // Martijn start - OBFHELPER
+    public ChunkMap _chunkMap() {
+        return this.chunkMap;
+    }
+    // Martijn end - OBFHELPER
     private final DimensionDataStorage dataStorage;
     private long lastInhabitedUpdate;
     public boolean spawnEnemies = true;
@@ -1044,6 +1054,14 @@ public class ServerChunkCache extends ChunkSource {
         this.distanceManager.updateSimulationDistance(simulationDistance);
     }
 
+    // Martijn start - OBFHELPER
+
+    public void _setSimulationDistance(int simulationDistance) {
+        this.setSimulationDistance(simulationDistance);
+    }
+
+    // Martijn end - OBFHELPER
+
     @Override
     public void setSpawnSettings(boolean spawnMonsters, boolean spawnAnimals) {
         this.spawnEnemies = spawnMonsters;
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 3f749f33f9168455cfe14a6e6120f36b412205f3..92901a6210296f97fd60248356a21960c84d7211 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -49,6 +49,7 @@ import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.SectionPos;
 import net.minecraft.core.particles.ParticleOptions;
+import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.protocol.Packet;
@@ -188,6 +189,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
     private static final int MAX_SCHEDULED_TICKS_PER_TICK = 65536;
     public final List<ServerPlayer> players;
     public final ServerChunkCache chunkSource;
+    // Martijn start - OBFHELPER
+    public ServerChunkCache _chunkSource() {
+        return this.chunkSource;
+    }
+    // Martijn end - OBFHELPER
     private final MinecraftServer server;
     public final PrimaryLevelData serverLevelData; // CraftBukkit - type
     final EntityTickList entityTickList;
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index bd1df088043427cb8d2d3e1fb60b32a007d9ca96..684c4bef872f416bf80fea01b680a6d86f3bf90d 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -476,6 +476,11 @@ public class ServerPlayer extends Player {
     private static final int NEUTRAL_MOB_DEATH_NOTIFICATION_RADII_XZ = 32;
     private static final int NEUTRAL_MOB_DEATH_NOTIFICATION_RADII_Y = 10;
     public ServerGamePacketListenerImpl connection;
+    // Martijn start - OBFHELPER
+    public ServerGamePacketListenerImpl _connection() {
+        return this.connection;
+    }
+    // Martijn end - OBFHELPER
     public net.minecraft.network.Connection networkManager; // Paper
     public final MinecraftServer server;
     public final ServerPlayerGameMode gameMode;
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 136c81ebb956f360d8828341164abf00dfcc212d..f31cc2bc3051f7bb0dcdc6fa1e582875a684690f 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -256,6 +256,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
     private static final int NO_BLOCK_UPDATES_TO_ACK = -1;
     private static final int PENDING_MESSAGE_DISCONNECT_THRESHOLD = 4096;
     public final Connection connection;
+    // Martijn start - OBFHELPER
+    public Connection _connection() {
+        return this.connection;
+    }
+    // Martijn end - OBFHELPER
     private final MinecraftServer server;
     public ServerPlayer player;
     private int tickCount;
@@ -2216,6 +2221,14 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         }
     }
 
+    // Martijn start - OBFHELPER
+
+    public void _send(Packet<?> packet, @Nullable PacketSendListener listener) {
+        this.send(packet, listener);
+    }
+
+    // Martijn end - OBFHELPER
+
     @Override
     public void handleSetCarriedItem(ServerboundSetCarriedItemPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.getLevel());
diff --git a/src/main/java/net/minecraft/world/entity/EntityType.java b/src/main/java/net/minecraft/world/entity/EntityType.java
index ec0319dd4b115e18b368027cc5dbe4d4d9c64840..fc2cb8713093610d48ad9cc191b70df61a47d5a5 100644
--- a/src/main/java/net/minecraft/world/entity/EntityType.java
+++ b/src/main/java/net/minecraft/world/entity/EntityType.java
@@ -19,6 +19,7 @@ import net.minecraft.core.Registry;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.nbt.Tag;
+import net.minecraft.network.Connection;
 import net.minecraft.network.chat.Component;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
@@ -261,6 +262,11 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T> {
     public static final EntityType<Turtle> TURTLE = EntityType.register("turtle", EntityType.Builder.of(Turtle::new, MobCategory.CREATURE).sized(1.2F, 0.4F).clientTrackingRange(10));
     public static final EntityType<Vex> VEX = EntityType.register("vex", EntityType.Builder.of(Vex::new, MobCategory.MONSTER).fireImmune().sized(0.4F, 0.8F).clientTrackingRange(8));
     public static final EntityType<Villager> VILLAGER = EntityType.register("villager", EntityType.Builder.of(Villager::new, MobCategory.MISC).sized(0.6F, 1.95F).clientTrackingRange(10));
+    // Martijn start - OBFHELPER
+    public static EntityType<Villager> _VILLAGER() {
+        return VILLAGER;
+    }
+    // Martijn end - OBFHELPER
     public static final EntityType<Vindicator> VINDICATOR = EntityType.register("vindicator", EntityType.Builder.of(Vindicator::new, MobCategory.MONSTER).sized(0.6F, 1.95F).clientTrackingRange(8));
     public static final EntityType<WanderingTrader> WANDERING_TRADER = EntityType.register("wandering_trader", EntityType.Builder.of(WanderingTrader::new, MobCategory.CREATURE).sized(0.6F, 1.95F).clientTrackingRange(10));
     public static final EntityType<Warden> WARDEN = EntityType.register("warden", EntityType.Builder.of(Warden::new, MobCategory.MONSTER).sized(0.9F, 2.9F).clientTrackingRange(16).fireImmune());
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index cd47b6882509888f7d34e95b4687ca7a9eab6caf..5a07f5e3119d82f8055e9d925f3f728b5afc1be2 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -140,6 +140,11 @@ public final class ItemStack {
     private Item item;
     @Nullable
     public CompoundTag tag;
+    // Martijn start - OBFHELPER
+    public @Nullable CompoundTag _tag() {
+        return this.tag;
+    }
+    // Martijn end - OBFHELPER
     private boolean emptyCacheFlag;
     @Nullable
     private Entity entityRepresentation;
