From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Tue, 20 Sep 2022 03:14:49 +0200
Subject: [PATCH] Configurable TPS catch-up time


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index d8485f1a5f4d3ca0554670cdf8e829514b1e7860..dc353b347fee6a5aa81ba8c39df7dd49bb1b0abb 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -445,6 +445,21 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     // Suki end - custom enchantability
 
+    // Suki start - configurable TPS catch-up time
+
+    public TpsCatchup tpsCatchup;
+
+    public class TpsCatchup extends ConfigurationPart {
+
+        /**
+         * In milliseconds (default 50)
+         */
+        public int millisFromNow = 50;
+
+    }
+
+    // Suki end - configurable TPS catch-up time
+
     public Messages messages;
 
     public class Messages extends ConfigurationPart {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 03694443ea1fde6718ed6acbccf81a5c6972e9a7..cf3389f045e4b6cc0f48888be75c509e74a3f94b 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -13,6 +13,7 @@ import com.mojang.authlib.GameProfileRepository;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.configuration.GlobalConfiguration;
 import it.unimi.dsi.fastutil.longs.LongIterator;
 import java.awt.image.BufferedImage;
 import java.io.BufferedWriter;
@@ -1205,7 +1206,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
                 lastTickTime = java.time.Duration.ofNanos(System.nanoTime() - tickStart); // Yatopia
                 this.profiler.popPush("nextTickWait");
                 this.mayHaveDelayedTasks = true;
-                this.delayedTasksMaxNextTickTime = Math.max(Util.getMillis() + 50L, this.nextTickTime);
+                this.delayedTasksMaxNextTickTime = Math.max(Util.getMillis() + GlobalConfiguration.get().tpsCatchup.millisFromNow, this.nextTickTime); // Suki - configurable TPS catch-up time
                 this.waitUntilNextTick();
                 this.profiler.pop();
                 this.endMetricsRecordingTick();
