From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Sun, 9 Oct 2022 17:22:02 +0200
Subject: [PATCH] Hide player game modes


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 7078bb109a6763d021d3cd002e094cc4ca9b35da..6af7bd59cf3bd8db29cb7a0b8196f1877536c5c7 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -9,8 +9,10 @@ import net.kyori.adventure.text.format.NamedTextColor;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ServerboundPlaceRecipePacket;
 import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.level.GameType;
 import org.bukkit.Bukkit; // Pufferfish
 import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.GameMode;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.spongepowered.configurate.objectmapping.ConfigSerializable;
 import org.spongepowered.configurate.objectmapping.meta.Comment;
@@ -18,6 +20,7 @@ import org.spongepowered.configurate.objectmapping.meta.Required;
 import org.spongepowered.configurate.objectmapping.meta.Setting;
 
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Objects;
 import java.util.logging.Level; // Pufferfish
@@ -387,6 +390,44 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     // Suki end - login protocol
 
+    // Suki start - hide game modes
+
+    public HideGameModes hideGameModes;
+
+    public class HideGameModes extends ConfigurationPart {
+
+        /**
+         * Any of the following (case-insensitive):
+         * <ul>
+         * <li>survival</li>
+         * <li>creative</li>
+         * <li>adventure</li>
+         * <li>spectator</li>
+         * </ul>
+         * Any other value disables this feature
+         */
+        public String replacement = "none";
+
+        public GameMode replace(GameMode original) {
+            try {
+                return GameMode.valueOf(replacement.toUpperCase(Locale.ROOT));
+            } catch (Exception ignored) {
+                return original;
+            }
+        }
+
+        public GameType replace(GameType original) {
+            try {
+                return GameType.valueOf(replacement.toUpperCase(Locale.ROOT));
+            } catch (Exception ignored) {
+                return original;
+            }
+        }
+
+    }
+
+    // Suki end - hide game modes
+
     public Messages messages;
 
     public class Messages extends ConfigurationPart {
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java
index 46ae5d40c26264e6b0ca253afa65b7846e6e5e6d..1bfb794a5780d649dc1bb5d6172f4b629614cb3d 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java
@@ -6,6 +6,8 @@ import com.mojang.authlib.GameProfile;
 import java.util.Collection;
 import java.util.List;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.configuration.GlobalConfiguration;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.protocol.Packet;
@@ -87,7 +89,7 @@ public class ClientboundPlayerInfoPacket implements Packet<ClientGamePacketListe
             @Override
             protected void write(FriendlyByteBuf buf, ClientboundPlayerInfoPacket.PlayerUpdate entry) {
                 buf.writeGameProfile(entry.getProfile());
-                buf.writeVarInt(entry.getGameMode().getId());
+                buf.writeVarInt(GlobalConfiguration.get().hideGameModes.replace(entry.getGameMode()).getId()); // Suki - hide game modes
                 buf.writeVarInt(entry.getLatency());
                 buf.writeNullable(entry.getDisplayName(), FriendlyByteBuf::writeComponent);
                 buf.writeNullable(entry.getProfilePublicKey(), (buf2, publicKeyData) -> {
@@ -106,7 +108,7 @@ public class ClientboundPlayerInfoPacket implements Packet<ClientGamePacketListe
             @Override
             protected void write(FriendlyByteBuf buf, ClientboundPlayerInfoPacket.PlayerUpdate entry) {
                 buf.writeUUID(entry.getProfile().getId());
-                buf.writeVarInt(entry.getGameMode().getId());
+                buf.writeVarInt(GlobalConfiguration.get().hideGameModes.replace(entry.getGameMode()).getId()); // Suki - hide game modes
             }
         },
         UPDATE_LATENCY {
