From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 10 Oct 2022 14:16:59 +0200
Subject: [PATCH] (WIP) Show vanished players in tab list


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 68d1722faf79afbda9d3cdeb8618dd34d82dacfb..a24c58a4a42d6948e869544af9ebfd5c775d03ab 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -437,6 +437,18 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     // Suki end - hide game modes
 
+    // Suki start - show vanished players in tab list
+
+    public VanishedPlayers vanishedPlayers;
+
+    public class VanishedPlayers extends ConfigurationPart {
+
+        public boolean showInTabList = false;
+
+    }
+
+    // Suki end - show vanished players in tab list
+
     public Messages messages;
 
     public class Messages extends ConfigurationPart {
