From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 10 Oct 2022 14:16:59 +0200
Subject: [PATCH] (WIP) Show vanished players in tab list


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 0891eb5a3cd87421d3a982429b887936f48bb0e1..e8ed948e91eeda3f76c081fda9195ec0ed9d40aa 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -528,6 +528,18 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     // Suki end - hide game modes
 
+    // Suki start - show vanished players in tab list
+
+    public VanishedPlayers vanishedPlayers;
+
+    public class VanishedPlayers extends ConfigurationPart {
+
+        public boolean showInTabList = false;
+
+    }
+
+    // Suki end - show vanished players in tab list
+
     public Messages messages;
 
     public class Messages extends ConfigurationPart {
