From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MartijnMuijsers <martijnmuijsers@live.nl>
Date: Mon, 10 Oct 2022 14:16:59 +0200
Subject: [PATCH] (WIP) Show vanished players in tab list


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 290d5f7efa7317a8405cffac7f22e466dac98b5c..9ee11d5f8c7dcd9439d12eca2f834e2c573d0903 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -534,6 +534,18 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     // Suki end - hide game modes
 
+    // Suki start - show vanished players in tab list
+
+    public VanishedPlayers vanishedPlayers;
+
+    public class VanishedPlayers extends ConfigurationPart {
+
+        public boolean showInTabList = false;
+
+    }
+
+    // Suki end - show vanished players in tab list
+
     public Messages messages;
 
     public class Messages extends ConfigurationPart {
